apps/dcos_dns/src/dcos_dns_handler.erl

```diff
@@ -177,90 +177,106 @@ resolve_loop(Begin, Workers, Fun, FailMsg, Zone) ->
+            {error, Reason} ->
+                lager:warning(
+                    "DNS worker [~p] ~p failed with ~p",
+                    [Protocol, UpstreamAddress, Reason]),
+                prometheus_counter:inc(
+                    dns, worker_failures_total,
+                    [Protocol, UpstreamAddress, Reason], 1)
+        end
     end).
 
--spec(init_worker(protocol(), pid(), upstream(), binary()) -> ok).
-init_worker(udp, Pid, Upstreams, Request) ->
-    udp_worker(Pid, Upstreams, Request);
-init_worker(tcp, Pid, Upstreams, Request) ->
-    tcp_worker(Pid, Upstreams, Request).
+-spec(init_worker(protocol(), pid(), upstream(), binary()) ->
+    {ok, binary()} | {error, atom()}).
+init_worker(udp, Pid, Upstream, Request) ->
+    udp_worker(Pid, Upstream, Request);
+init_worker(tcp, Pid, Upstream, Request) ->
+    tcp_worker(Pid, Upstream, Request).
 
--spec(udp_worker(pid(), upstream(), binary()) -> ok).
-udp_worker(Pid, Upstream = {IP, Port}, Request) ->
-    StartTime = erlang:monotonic_time(millisecond),
+-spec(udp_worker(pid(), upstream(), binary()) ->
+    {ok, binary()} | {error, atom()}).
+udp_worker(Pid, {IP, Port}, Request) ->
     Opts = [{reuseaddr, true}, {active, once}, binary],
     case gen_udp:open(0, Opts) of
         {ok, Socket} ->
             case gen_udp:send(Socket, IP, Port, Request) of
                 ok ->
-                    udp_worker(StartTime, Pid, Socket, Upstream);
-                {error, Error} ->
-                    lager:warning("DNS worker ~p failed with ~p", [Upstream, Error])
+                    udp_worker_receive(Pid, {IP, Port}, Socket);
+                {error, Reason} ->
+                    {error, Reason}
             end;
-        {error, Error} ->
-            lager:warning("DNS worker ~p failed with ~p", [Upstream, Error])
+        {error, Reason} ->
+            {error, Reason}
     end.
 
--spec(udp_worker(integer(), pid(), gen_udp:socket(), upstream()) -> ok).
-udp_worker(StartTime, Pid, Socket, Upstream = {IP, Port}) ->
+-spec(udp_worker_receive(pid(), upstream(), gen_udp:socket()) ->
+    {ok, binary()} | {error, atom()}).
+udp_worker_receive(Pid, {IP, Port}, Socket) ->
     MonRef = erlang:monitor(process, Pid),
     try
         receive
-            {'DOWN', MonRef, process, _Pid, normal} ->
-                ok;
-            {'DOWN', MonRef, process, _Pid, Reason} ->
-                exit(Reason);
+            {'DOWN', MonRef, process, _Pid, _Reason} ->
+                {error, down};
             {udp, Socket, IP, Port, Response} ->
-                Pid ! {reply, self(), Response},
-                report_latency([?MODULE, Upstream, latency], StartTime);
+                {ok, Response};
             {timeout, Pid} ->
-                lager:warning("DNS worker ~p timed out", [Upstream])
+                {error, timeout}
         after ?TIMEOUT ->
-            ok
+            {error, timeout}
         end
     after
         gen_udp:close(Socket)
```

Comment:
With new semantics, I'd recommend to move `gen_udp:close/1` call to `udp_worker/3` function.